torch/csrc/jit/codegen/cuda/compute_at.cpp

```diff
@@ -269,37 +229,58 @@ unsigned int ComputeAt::backwardComputeAt_impl(
 unsigned int ComputeAt::forwardComputeAt_impl(
     TensorView* producer,
     TensorView* consumer,
-    unsigned int producer_compute_at_axis) {
+    unsigned int producer_compute_at_pos) {
   FUSER_PERF_SCOPE("forwardComputeAt_impl");
 
+  // Can get into a situation where we inlined into a reduction, but then would
+  // try to traverse forward at that position but wouldn't be valid.
+  // Reduce position to be inside first reduction
+  unsigned int first_red_pos = producer->nDims();
+  for (unsigned int i = 0;
+       i < (unsigned int)producer->domain()->domain().size();
+       i++) {
+    if (producer->axis(i)->isReduction()) {
+      first_red_pos = i;
+      break;
+    }
+  }
+  producer_compute_at_pos = std::min(first_red_pos, producer_compute_at_pos);
+  if (producer_compute_at_pos == 0) {
+    return 0;
+  }
+
   auto& consumer_entry = tv_data.at(consumer);
   const auto& producer_entry = tv_data.at(producer);
 
   auto replay = TransformReplay::replayCasP(
       consumer->domain(),
       producer->domain(),
-      (int)producer_compute_at_axis,
+      (int)producer_compute_at_pos,
       root_map_);
 
-  if (producer_entry.shouldSetComputeAt(producer_compute_at_axis)) {
-    int producer_rel_pos = replay.second;
-    int producer_this_pos = (int)producer_compute_at_axis;
-    // When the producer CA axes have reductions, they are not used to
-    // replay the consumer.
-    if (producer_this_pos > producer_rel_pos) {
-      producer_this_pos = producer_rel_pos;
-    }
-    producer->setComputeAt(producer_this_pos);
-  }
-
   consumer_entry.setPassPosition(replay.second);
-  if (consumer_entry.shouldSetComputeAt(replay.second) &&
-      !(consumer == consumer_ && reference_ == consumer_)) {
-    const TensorDomain* current_domain = consumer->domain();
-    TensorDomain* new_domain = replay.first;
-    consumer->setDomain(new_domain);
-    root_map_.setAlias(current_domain, new_domain);
-    consumer_entry.setComputeAtDomain(consumer->domain());
+
+  // Should set compute at isn't quite right at the moment, still important for
+  // the consumer usage though. Need to refactor.
+  if (producer_compute_at_pos >= producer->getComputeAtPosition()) {
+    producer->setComputeAt((int)producer_compute_at_pos);
+
+    if (replay.second > consumer->getMaxProducerPosition()) {
+      const TensorDomain* current_domain = consumer->domain();
+      TensorDomain* new_domain = replay.first;
+
+      TORCH_INTERNAL_ASSERT(
+          validateDomain(consumer, new_domain),
+          "Tried to set the domain of ",
+          producer,
+          " to ",
+          new_domain,
+          " but that would invalidate previously compute at position or max producer position.");
+
+      consumer->setDomain(new_domain);
+      consumer->setMaxProducer(replay.second);
+      root_map_.setAlias(current_domain, new_domain);
+    }
```

Comment:
Why this part of code needs to be predicated by `producer_compute_at_pos >= producer->getComputeAtPosition()`?

Reply:
In the old version:
```  if (producer_entry.shouldSetComputeAt(producer_compute_at_axis)) {```
and
```  if (consumer_entry.shouldSetComputeAt(replay.second) &&
      !(consumer == consumer_ && reference_ == consumer_)) ```
were always identical. It's the reason we could remove `shouldSetComputeAt`

Reply:
In the old version, it was guarded by:

```
if (consumer_entry.shouldSetComputeAt(replay.second) &&	
      !(consumer == consumer_ && reference_ == consumer_)) {
```
https://github.com/csarofeen/pytorch/pull/683/files#diff-a09f5a841235343227ca4aa5021932ff8ef7c883578679bb366652ba0c18884cL296-L297

In the new version, it is guarded by:

```
 if (replay.second > consumer->getMaxProducerPosition()) {
```

*AND*

```
if (producer_compute_at_pos >= producer->getComputeAtPosition()) {
```

Is the second one needed?